src/solc_0.8/ReferralValidator/ReferralValidator08.sol

```diff
@@ -0,0 +1,186 @@
+        }
+
+        destination.transfer(amountForDestination);
+    }
+
+    function handleReferralWithERC20(
+        address buyer,
+        uint256 amount,
+        bytes memory referral,
+        address payable destination,
+        address tokenAddress
+    ) internal {
+        ERC20Token token = ERC20Token(tokenAddress);
+        uint256 amountForDestination = amount;
+
+        if (referral.length > 0) {
+            (bytes memory signature, address referrer, address referee, uint256 expiryTime, uint256 commissionRate) =
+                decodeReferral(referral);
+
+            uint256 commission = 0;
+
+            if (isReferralValid(signature, referrer, referee, expiryTime, commissionRate)) {
+                commission = SafeMath.div(SafeMath.mul(amount, commissionRate), 10000);
+
+                emit ReferralUsed(referrer, referee, tokenAddress, amount, commission, commissionRate);
+                amountForDestination = SafeMath.sub(amountForDestination, commission);
+            }
+
+            if (commission > 0) {
+                require(token.transferFrom(buyer, referrer, commission), "commision transfer failed");
+            }
+        }
+
+        require(token.transferFrom(buyer, destination, amountForDestination), "payment transfer failed");
+    }
+
+    /**
+     * @notice Check if a referral is valid
+     * @param signature The signature to check (signed referral)
+     * @param referrer The address of the referrer
+     * @param referee The address of the referee
+     * @param expiryTime The expiry time of the referral
+     * @param commissionRate The commissionRate of the referral
+     * @return True if the referral is valid
+     */
+    function isReferralValid(
+        bytes memory signature,
+        address referrer,
+        address referee,
+        uint256 expiryTime,
+        uint256 commissionRate
+    ) public view returns (bool) {
+        if (commissionRate > _maxCommissionRate || referrer == referee || block.timestamp > expiryTime) {
+            return false;
+        }
+
+        bytes32 hashedData = keccak256(abi.encodePacked(referrer, referee, expiryTime, commissionRate));
+
+        address signer =
+            ECDSA.recover(keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", hashedData)), signature);
+
+        if (_previousSigningWallets[signer] >= block.timestamp) {
+            return true;
+        }
+
+        return _signingWallet == signer;
+    }
+
+    function decodeReferral(bytes memory referral)
+        public
+        pure
+        returns (
+            bytes memory,
+            address,
+            address,
+            uint256,
+            uint256
+        )
+    {
+        (bytes memory signature, address referrer, address referee, uint256 expiryTime, uint256 commissionRate) =
+            abi.decode(referral, (bytes, address, address, uint256, uint256));
+
+        return (signature, referrer, referee, expiryTime, commissionRate);
+    }
+}
```

Comment:
Add tests for this class

Reply:
done -> https://github.com/thesandboxgame/sandbox-smart-contracts/pull/413/commits/9f20c792a01a98e192c28eece98f3511209dbac7